crates/red_knot_python_semantic/src/types.rs

```diff
@@ -1,91 +1,70 @@
 use ruff_db::files::File;
-use ruff_db::parsed::parsed_module;
 use ruff_python_ast::name::Name;
 
-use crate::semantic_index::symbol::{NodeWithScopeKind, PublicSymbolId, ScopeId};
-use crate::semantic_index::{public_symbol, root_scope, semantic_index, symbol_table};
-use crate::types::infer::{TypeInference, TypeInferenceBuilder};
+use crate::semantic_index::definition::Definition;
+use crate::semantic_index::symbol::{ScopeId, ScopedSymbolId};
+use crate::semantic_index::{root_scope, symbol_table, use_def_map};
 use crate::{Db, FxOrderSet};
 
 mod display;
 mod infer;
 
-/// Infers the type of a public symbol.
-///
-/// This is a Salsa query to get symbol-level invalidation instead of file-level dependency invalidation.
-/// Without this being a query, changing any public type of a module would invalidate the type inference
-/// for the module scope of its dependents and the transitive dependents because.
-///
-/// For example if we have
-/// ```python
-/// # a.py
-/// import x from b
-///
-/// # b.py
-///
-/// x = 20
-/// ```
-///
-/// And x is now changed from `x = 20` to `x = 30`. The following happens:
-///
-/// * The module level types of `b.py` change because `x` now is a `Literal[30]`.
-/// * The module level types of `a.py` change because the imported symbol `x` now has a `Literal[30]` type
-/// * The module level types of any dependents of `a.py` change because the imported symbol `x` now has a `Literal[30]` type
-/// * And so on for all transitive dependencies.
-///
-/// This being a query ensures that the invalidation short-circuits if the type of this symbol didn't change.
-#[salsa::tracked]
-pub(crate) fn public_symbol_ty<'db>(db: &'db dyn Db, symbol: PublicSymbolId<'db>) -> Type<'db> {
-    let _span = tracing::trace_span!("public_symbol_ty", ?symbol).entered();
-
-    let file = symbol.file(db);
-    let scope = root_scope(db, file);
-
-    // TODO switch to inferring just the definition(s), not the whole scope
-    let inference = infer_types(db, scope);
-    inference.symbol_ty(symbol.scoped_symbol_id(db))
-}
+pub(crate) use self::infer::{infer_definition_types, infer_expression_types, infer_scope_types};
 
-/// Shorthand for `public_symbol_ty` that takes a symbol name instead of a [`PublicSymbolId`].
-pub(crate) fn public_symbol_ty_by_name<'db>(
+/// Infer the public type of a symbol (its type as seen from outside its scope).
+pub(crate) fn symbol_ty<'db>(
     db: &'db dyn Db,
-    file: File,
-    name: &str,
-) -> Option<Type<'db>> {
-    let symbol = public_symbol(db, file, name)?;
-    Some(public_symbol_ty(db, symbol))
-}
+    scope: ScopeId<'db>,
+    symbol: ScopedSymbolId,
+) -> Type<'db> {
+    let _span = tracing::trace_span!("symbol_ty", ?symbol).entered();
+
+    let usedef = use_def_map(db, scope);
+    let mut def_types = usedef
+        .public_definitions(symbol)
+        .iter()
+        .map(|def| definition_ty(db, *def));
+
+    let Some(first) = def_types.next() else {
+        return Type::Unbound;
+    };
```

Comment:
```suggestion
   let first = def_types.next().unwrap_or(Type::Unbound);
```

Reply:
Is this better even though it means we'll make an entirely unnecessary second call to `def_types.next()` that we already know will return `None`? It seems like we may as well short-circuit here and just return Unbound directly.